### DISCUSSION TITLE
Devin / Devika implementation?

### DISCUSSION THREAD

**Discussion by olimorris** [olimorris/codecompanion.nvim#31] [Ideas]
I'm mulling over the idea of adding a [Devin](https://www.cognition-labs.com/introducing-devin) / [Devika](https://github.com/stitionai/devika) type feature to the plugin.

With the Neovim ecosystem right now we have the joy of fantastic [code runners](https://github.com/stevearc/overseer.nvim) and access to an in-built terminal. I guess the missing piece would be an "agent" which connects everything together.

Keen to get people's thoughts on this and how it **could** look from a workflow perspective, ideally from inside a Neovim buffer.

**Comment by lazymaniac** [comment]
Something like https://docs.crewai.com could help. It supports multiple agents, and tools for agents like webserchaing, file reading etc. Short video: https://www.youtube.com/watch?v=s-jYxgKMqRc

**Reply by olimorris** [reply]
Great resource, thank you. Microsoft's [autogen](https://microsoft.github.io/autogen/docs/tutorial/introduction/) is something I've been looking at too. Both appear to be great choices.

I'm mulling over how this feature could be implemented in CodeCompanion. I think the limiting factor right now is a bridge between CodeCompanion and the AI Agents. Ultimately, my desire would be to connect the plugin to them to display output, progress and ultimately put the end result in a buffer.

However both only support LLMs that follow OpenAI's API.

**Reply by lazymaniac** [reply]
Ok, a third option might be [https://python.langchain.com/docs/get_started/introduction](url). CrewAi is built on top of LangChain. It offers a few frameworks where LangChain serves as the base, supporting prompt templates, various LLM providers, embeddings, vector stores, retrievers, memory, and dozens of tools. Additionally, there is LangGraph, which allows the construction of a graph of interactions between agents. I think LangChain currently has the most built-in integrations with anything. However, for someone like me who isn't a Python coder, it's kind of tricky and complicated to use. I'm currently trying to build agents with LangChain and LangGraph that can read project files and vectorize them as a knowledge base, and then add additional knowledge retrievers based on web browsing.

**Reply by olimorris** [reply]
Great shout!! Completely overlooked LangChain

**Comment by nuvic** [comment]
I recently watched Andrew Ng's video on agentic workflows and this could be something we could do here: https://www.youtube.com/watch?v=sal78ACtGTc

One simple way is to have a workflow for self-reflection:
- user prompts to create a function to do xyz
- enter agentic workflow, prompt an agent to come up with a plan for the function
- have another agent (or even itself) critique that plan
- use the initial plan and critique to come up with a better plan
- use the better plan to come up with the code for the function

An agentic workflow for tool use could be interesting tool (using the code runner linked above)

**Reply by nuvic** [reply]
oh just realized that you created a PR already for this. awesome!

**Reply by olimorris** [reply]
Could you try this out and let me know your thoughts? Happy to tweak the default prompts in the plugin.

**Reply by nuvic** [reply]
Awesome! I tried it out, and it works great so far. the warning on the readme is also useful, since it outputs quite a bit of tokens.
i added some comments about the default prompts in the PR

**Reply by olimorris** [reply]
Just seen them. Some brilliant suggestions and thank you.

The token use actually got me exploring Ollama a bit more and I've made some changes to the default model on the adapter. That is offering some really good output so far.

**Reply by nuvic** [reply]
Oh yes I also saw the change in default model, thank you for that! I don't have a GPU so I've been using anthropic. Switching to haiku is much more cost efficient =)

**Reply by olimorris** [reply]
I found the same! So much faster too.

**Comment by olimorris** [comment]
Hi All.

Coming back to this, I wanted to share the latest work that I've been doing with the chat buffer in the form of variables and agents. I've posted about it here: https://github.com/olimorris/codecompanion.nvim/discussions/76.

Would love to get your feedback
